### PR TITLE
#62 - Storage.list() method accepts Key as prefix

### PR DESCRIPTION
**Pull request by olegmoz**
Issue: #62 
This PR modifies Storage interface, FileStorage implementation and related classes such as BlockingStorage, RxStorage etc

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#63](https://github.com/artipie/asto/issues/63) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/c1189f83-8f79-4c5e-8f01-6a254345ba3c); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 1]
```diff
@@ -109,4 +111,22 @@ void move(@TempDir final Path tmp) {
         storage.move(source, destination);
         MatcherAssert.assertThat(storage.value(destination), Matchers.equalTo(data));
     }
+
+    @Test
+    void list(@TempDir final Path tmp) {
```
Comment (iakunin) [inline src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz let's also test some negative cases when no objects found.

Comment (olegmoz) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @iakunin nice point! Added the test and changed behavior of list() method in this case, so it doesn't crash, but returns empty collections. Think it's better. Thank you!

Comment (iakunin) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz thank you for your responsiveness =)

**Comment by olegmoz** [comment]
@iakunin fixed two bugs and fixed the tests, please review

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 2]
```diff
@@ -72,27 +74,24 @@ public FileStorage(final Path path) {
...
+                final Collection<Key> keys;
+                if (Files.exists(path)) {
+                    final int dirnamelen = path.toString().length() - prefix.string().length();
+                    keys = Files.walk(path)
+                        .filter(Files::isRegularFile)
+                        .map(Path::toString)
+                        .map(p -> p.substring(dirnamelen))
+                        .map(s -> s.split(FileSystems.getDefault().getSeparator()))
+                        .map(Key.From::new)
+                        .sorted(Comparator.comparing(Key.From::string))
```
Comment (iakunin) [inline src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz could you please give me a hint why should we sort keys here? It's not so obvious for me. Or even better write a test instead of the explanation - that would be the best explanation =)

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 3]
```diff
@@ -72,27 +74,24 @@ public FileStorage(final Path path) {
...
-                    );
+                final Path path = this.path(prefix);
+                final Collection<Key> keys;
+                if (Files.exists(path)) {
+                    final int dirnamelen = path.toString().length() - prefix.string().length();
+                    keys = Files.walk(path)
+                        .filter(Files::isRegularFile)
+                        .map(Path::toString)
+                        .map(p -> p.substring(dirnamelen))
+                        .map(s -> s.split(FileSystems.getDefault().getSeparator()))
```
Comment (iakunin) [inline src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz seems that now the tests could fail on the wrong filesystem (on Windows, for example). Can we prevent this somehow? Or write system-dependent tests (but I don't like this idea). It's better to write system independent tests.

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 2 above)
Comment (olegmoz) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @iakunin actually existing test checks order, because compares output with list, not a set. I added sorting, to make output independent from underlying OS. It happens that Unix and Windows do `Files.walk` in different order, maybe it also depends on file system. I think that it is a good idea, to guarantee order of keys regardless of OS/FS

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 3 above)
Comment (olegmoz) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @iakunin on the contrary, before my last commits it failed on Unix. Now tests run on both Unix and Windows, see the checks - we run tests on both systems.

**Comment by olegmoz** [comment]
@iakunin hope I have answered your questions, please review

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 2 above)
Comment (iakunin) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz got it. Now it's much clearer for me, why it was done.

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 3 above)
Comment (iakunin) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz I see. Now it's much clearer for me, why it was done. Thanks for your explanation.

**Comment by olegmoz** [comment]
@g4s8 please review

**Comment by olegmoz** [comment]
@iakunin please re-review, there was trivial change on merge of master branch

**Comment by olegmoz** [comment]
@g4s8 please review

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 4]
```diff
@@ -103,7 +103,11 @@ public FileStorage(final Path path, final FileSystem fls) {
                         .filter(Files::isRegularFile)
                         .map(Path::toString)
                         .map(p -> p.substring(dirnamelen))
-                        .map(s -> s.split(FileSystems.getDefault().getSeparator()))
+                        .map(
+                            s -> s.split(
+                                FileSystems.getDefault().getSeparator().replace("\\", "\\\\")
```
Comment (iakunin) [inline src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz could you give me the case, when replacement is needed. Cannot come up with it.

Comment (olegmoz) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz sure! Thing is `String.split()` method accepts regex as parameter, not just string. When file system separator is `\` then it's illegal regex and has to be converted into '\\'. That is why I needed to add the replacement.

**Comment by 0crat** [comment]
@iakunin[/z](https://www.0crat.com/u/iakunin) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 4 above)
Comment (iakunin) [reply src/main/java/com/artipie/asto/fs/FileStorage.java]: @olegmoz now it's clear for me, thank you so much!

**Comment by iakunin** [comment]
@olegmoz could you resolve conflicts with master branch, please?

**Comment by olegmoz** [comment]
@iakunin merged with master, please review

**Comment by olegmoz** [comment]
@g4s8 please merge

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @iakunin[/z](https://www.0crat.com/u/iakunin)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)
